Src/DependencyCollector/Shared.Tests/Implementation/SqlClientDiagnosticSourceListenerTests.cs

```diff
@@ -0,0 +1,434 @@
+
+            var parentActivity = new Activity("Parent");
+            var activity = new Activity("Current").AddBaggage("Stuff", "123");
+
+            try
+            {
+                parentActivity.Start();
+                activity.Start();
+
+                this.fakeSqlClientDiagnosticSource.Write(
+                    SqlClientDiagnosticSourceListener.SqlAfterExecuteCommand,
+                    afterExecuteEventData);
+
+                var dependencyTelemetry = (DependencyTelemetry)this.sendItems.Single();
+
+                Assert.AreEqual(activity.RootId, dependencyTelemetry.Context.Operation.Id);
+                Assert.AreEqual(parentActivity.Id, dependencyTelemetry.Context.Operation.ParentId);
+                Assert.AreEqual("123", dependencyTelemetry.Context.Properties["Stuff"]);
+            }
+            finally
+            {
+                activity.Stop();
+                parentActivity.Stop();
+            }
+        }
+
+        [TestMethod]
+        public void TracksCommandExecuted()
+        {
+            var operationId = Guid.NewGuid();
+            var sqlConnection = new SqlConnection(TestConnectionString);
+            var sqlCommand = sqlConnection.CreateCommand();
+            sqlCommand.CommandText = "select * from orders";
+            
+            var beforeExecuteEventData = new
+            {
+                OperationId = operationId,
+                Timestamp = 1000000L
+            };
+
+            this.fakeSqlClientDiagnosticSource.Write(
+                SqlClientDiagnosticSourceListener.SqlBeforeExecuteCommand,
+                beforeExecuteEventData);
+
+            var afterExecuteEventData = new
+            {
+                OperationId = operationId,
+                Operation = "ExecuteReader",
+                ConnectionId = sqlConnection.ClientConnectionId,
+                Command = sqlCommand,
+                Statistics = sqlConnection.RetrieveStatistics(),
+                Timestamp = 2000000L
+            };
+
+            var now = DateTimeOffset.UtcNow;
+
+            this.fakeSqlClientDiagnosticSource.Write(
+                SqlClientDiagnosticSourceListener.SqlAfterExecuteCommand,
+                afterExecuteEventData);
+
+            var dependencyTelemetry = (DependencyTelemetry)this.sendItems.Single();
+
+            Assert.AreEqual(afterExecuteEventData.OperationId.ToString("N"), dependencyTelemetry.Id);
+            Assert.AreEqual(sqlCommand.CommandText, dependencyTelemetry.Data);
+            Assert.AreEqual("(localdb)\\MSSQLLocalDB | master", dependencyTelemetry.Name);
+            Assert.AreEqual("(localdb)\\MSSQLLocalDB | master", dependencyTelemetry.Target);
+            Assert.AreEqual(RemoteDependencyConstants.SQL, dependencyTelemetry.Type);
+            Assert.IsTrue((bool)dependencyTelemetry.Success);
+            Assert.AreEqual(1000000L, dependencyTelemetry.Duration.Ticks);
+            Assert.IsTrue(dependencyTelemetry.Timestamp.Ticks < now.Ticks);
```

Comment:
The timestamp should be exactly the timestamp set inside beforeExecuteEventData.

Reply:
We can't because the timestamp in beforeExecuteEventData is the ticks returned by Stopwatch.GetTimestamp() whereas the DependencyTelemetry.Timestamp is DateTimeOffset. This is why we compute the value by subtracting duration from DateTimeOffset.UtcNow.

Reply:
We are not storing the start timestamp, but the ticks value returned by Stopwatch.GetTimestamp().

Reply:
ok got it